## Упражнения по библиотеке Numpy

In [98]:
import numpy as np

**1.** Дан случайный массив, поменять знак у элементов, значения которых между 3 и 8

In [99]:
arr = np.random.randint(0, 10, size=10)
print("Исходный массив:", arr)

arr[(arr > 3) & (arr < 8)] *= -1
print("Результат:", arr)

Исходный массив: [8 0 6 7 6 4 2 2 6 2]
Результат: [ 8  0 -6 -7 -6 -4  2  2 -6  2]


**2.** Заменить максимальный элемент случайного массива на 0

In [100]:
arr = np.random.randint(0, 100, size=10)
print("Исходный массив:", arr)

arr[arr.argmax()] = 0
print("Результат:", arr)

Исходный массив: [59 68 45 99 55 49 73 55 33 61]
Результат: [59 68 45  0 55 49 73 55 33 61]


**3.** Построить прямое произведение массивов (все комбинации с каждым элементом). На вход подается двумерный массив

In [101]:
# Исходный массив
arr = np.array([[1, 2],
                [3, 4],
                [5, 6]])

n = arr.shape[0]

# В одну строку
result = np.hstack([np.repeat(arr, n, axis=0), np.tile(arr, (n, 1))])

print(result)

[[1 2 1 2]
 [1 2 3 4]
 [1 2 5 6]
 [3 4 1 2]
 [3 4 3 4]
 [3 4 5 6]
 [5 6 1 2]
 [5 6 3 4]
 [5 6 5 6]]


**4.** Даны 2 массива A (8x3) и B (2x2). Найти строки в A, которые содержат элементы из каждой строки в B, независимо от порядка элементов в B

In [102]:
A = np.random.randint(1, 5, size=(8, 3))
B = np.random.randint(1, 5, size=(2, 2))

print("A:\n", A)
print("\nB:\n", B)

result = [i for i, row in enumerate(A)
          if all(set(b).issubset(set(row)) for b in B)]

print(f"\nИндексы строк: {result}")
print(f"Строки:\n{A[result]}")

A:
 [[2 3 2]
 [3 2 3]
 [1 3 3]
 [4 1 3]
 [4 3 3]
 [1 1 4]
 [3 4 2]
 [4 4 2]]

B:
 [[3 3]
 [2 4]]

Индексы строк: [6]
Строки:
[[3 4 2]]


**5.** Дана 10x3 матрица, найти строки из неравных значений (например строка [2,2,3] остается, строка [3,3,3] удаляется)

In [103]:
M = np.random.randint(1, 3, size=(10, 3))
print("Исходная:\n", M)

result = M[np.array([len(set(row)) > 1 for row in M])]
print("\nРезультат:\n", result)

Исходная:
 [[2 1 2]
 [2 2 1]
 [2 1 1]
 [1 1 2]
 [1 1 1]
 [2 1 2]
 [1 2 2]
 [1 1 2]
 [2 1 2]
 [1 1 2]]

Результат:
 [[2 1 2]
 [2 2 1]
 [2 1 1]
 [1 1 2]
 [2 1 2]
 [1 2 2]
 [1 1 2]
 [2 1 2]
 [1 1 2]]


**6.** Дан двумерный массив. Удалить те строки, которые повторяются

In [104]:
M = np.random.randint(1, 4, size=(8, 3))
print("Исходная матрица:\n", M)

result = np.unique(M, axis=0)
print("\nБез повторяющихся строк:\n", result)

Исходная матрица:
 [[2 1 3]
 [3 2 1]
 [3 1 2]
 [2 2 2]
 [1 2 1]
 [2 3 3]
 [1 1 2]
 [1 1 2]]

Без повторяющихся строк:
 [[1 1 2]
 [1 2 1]
 [2 1 3]
 [2 2 2]
 [2 3 3]
 [3 1 2]
 [3 2 1]]


______
______

Для каждой из следующих задач (1-5) нужно привести 2 реализации – одна без использования numpy (cчитайте, что там, где на входе или выходе должны быть numpy array, будут просто списки), а вторая полностью векторизованная с использованием numpy (без использования питоновских циклов/map/list comprehension).


__Замечание 1.__ Можно считать, что все указанные объекты непустые (к примеру, в __задаче 1__ на диагонали матрицы есть ненулевые элементы).

__Замечание 2.__ Для большинства задач решение занимает не больше 1-2 строк.

___

* __Задача 1__: Подсчитать произведение ненулевых элементов на диагонали прямоугольной матрицы.  
 Например, для X = np.array([[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]) ответ 3.

In [105]:
X = [[1, 0, 1], [2, 0, 2], [3, 0, 3], [4, 4, 4]]

# Без numpy
product = 1
min_dim = min(len(X), len(X[0]))
for i in range(min_dim):
    if X[i][i] != 0:
        product *= X[i][i]
print(product)

# С numpy
diag = np.diag(X)
result = np.prod(diag[diag != 0])
print(result)

3
3


* __Задача 2__: Даны два вектора x и y. Проверить, задают ли они одно и то же мультимножество.  
  Например, для x = np.array([1, 2, 2, 4]), y = np.array([4, 2, 1, 2]) ответ True.

In [106]:
# Без numpy
x = [1, 2, 2, 4]
y = [4, 2, 1, 2]
result = sorted(x) == sorted(y)
print(result)

# С numpy
x = np.array([1, 2, 2, 4])
y = np.array([4, 2, 1, 2])
result = np.array_equal(np.sort(x), np.sort(y))
print(result)

True
True


* __Задача 3__: Найти максимальный элемент в векторе x среди элементов, перед которыми стоит ноль.
 Например, для x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0]) ответ 5.

In [107]:
# Без numpy
x = [6, 2, 0, 3, 0, 0, 5, 7, 0]

max_val = None
for i in range(1, len(x)):
    if x[i-1] == 0:
        if max_val is None or x[i] > max_val:
            max_val = x[i]
print(max_val)

# С numpy
x = np.array([6, 2, 0, 3, 0, 0, 5, 7, 0])

mask = np.roll(x, 1) == 0
mask[0] = False
result = np.max(x[mask]) if np.any(mask) else None
print(result)

5
5


* __Задача 4__: Реализовать кодирование длин серий (Run-length encoding). Для некоторого вектора x необходимо вернуть кортеж из двух векторов одинаковой длины. Первый содержит числа, а второй - сколько раз их нужно повторить.  
 Например, для x = np.array([2, 2, 2, 3, 3, 3, 5]) ответ (np.array([2, 3, 5]), np.array([3, 3, 1])).

In [108]:
# Без numpy
x = [2, 2, 2, 3, 3, 3, 5]

values = []
counts = []
current_val = x[0]
count = 1
for i in range(1, len(x)):
    if x[i] == current_val:
        count += 1
    else:
        values.append(current_val)
        counts.append(count)
        current_val = x[i]
        count = 1
values.append(current_val)
counts.append(count)
print(values, counts)



# С numpy
x = np.array([2, 2, 2, 3, 3, 3, 5])

change_indices = np.where(x[:-1] != x[1:])[0] + 1
values = x[np.concatenate(([0], change_indices))]
starts = np.concatenate(([0], change_indices))
ends = np.concatenate((change_indices, [len(x)]))
counts = ends - starts
print(values, counts)

[2, 3, 5] [3, 3, 1]
[2 3 5] [3 3 1]


* __Задача 5__: Даны две выборки объектов - X и Y. Вычислить матрицу евклидовых расстояний между объектами. Сравните с функцией scipy.spatial.distance.cdist по скорости работы.

In [109]:
# Без numpy
X = [[1, 2], [3, 4], [5, 6]]
Y = [[1, 1], [2, 2], [3, 3]]

import math
n, m = len(X), len(Y)
distances = [[0] * m for _ in range(n)]
for i in range(n):
    for j in range(m):
        sum_sq = 0
        for k in range(len(X[i])):
            sum_sq += (X[i][k] - Y[j][k]) ** 2
        distances[i][j] = math.sqrt(sum_sq)
for row in distances:
    print(row)


# С numpy
X = np.array([[1, 2], [3, 4], [5, 6]])
Y = np.array([[1, 1], [2, 2], [3, 3]])

X_sq = np.sum(X**2, axis=1).reshape(-1, 1)
Y_sq = np.sum(Y**2, axis=1).reshape(1, -1)
distances = np.sqrt(X_sq + Y_sq - 2 * np.dot(X, Y.T))
print(distances)

[1.0, 1.0, 2.23606797749979]
[3.605551275463989, 2.23606797749979, 1.0]
[6.4031242374328485, 5.0, 3.605551275463989]
[[1.         1.         2.23606798]
 [3.60555128 2.23606798 1.        ]
 [6.40312424 5.         3.60555128]]


In [110]:
import scipy.spatial.distance as distance
import time

X = np.random.rand(1000, 50)
Y = np.random.rand(2000, 50)

start = time.time()
X_sq = np.sum(X**2, axis=1).reshape(-1, 1)
Y_sq = np.sum(Y**2, axis=1).reshape(1, -1)
distances = np.sqrt(X_sq + Y_sq - 2 * np.dot(X, Y.T))
time_custom = time.time() - start
print(f"Время нашей реализации: {time_custom:.4f} сек")

# scipy.spatial.distance.cdist
start = time.time()
dist_scipy = distance.cdist(X, Y, metric='euclidean')
time_scipy = time.time() - start
print(f"Время scipy.cdist: {time_scipy:.4f} сек")

Время нашей реализации: 0.0310 сек
Время scipy.cdist: 0.0883 сек


_______
________

* #### __Задача 6__: CrunchieMunchies __*__

Вы работаете в отделе маркетинга пищевой компании MyCrunch, которая разрабатывает новый вид вкусных, полезных злаков под названием **CrunchieMunchies**.

Вы хотите продемонстрировать потребителям, насколько полезны ваши хлопья по сравнению с другими ведущими брендами, поэтому вы собрали данные о питании нескольких разных конкурентов.

Ваша задача - использовать вычисления Numpy для анализа этих данных и доказать, что ваши **СrunchieMunchies** - самый здоровый выбор для потребителей.


In [111]:
import numpy as np

1. Просмотрите файл cereal.csv. Этот файл содержит количества калорий для различных марок хлопьев. Загрузите данные из файла и сохраните их как calorie_stats.

In [112]:
calorie_stats = np.loadtxt("./data/cereal.csv", delimiter=",")
calorie_stats

array([ 70., 120.,  70.,  50., 110., 110., 110., 130.,  90.,  90., 120.,
       110., 120., 110., 110., 110., 100., 110., 110., 110., 100., 110.,
       100., 100., 110., 110., 100., 120., 120., 110., 100., 110., 100.,
       110., 120., 120., 110., 110., 110., 140., 110., 100., 110., 100.,
       150., 150., 160., 100., 120., 140.,  90., 130., 120., 100.,  50.,
        50., 100., 100., 120., 100.,  90., 110., 110.,  80.,  90.,  90.,
       110., 110.,  90., 110., 140., 100., 110., 110., 100., 100., 110.])

2. В одной порции CrunchieMunchies содержится 60 калорий. Насколько выше среднее количество калорий у ваших конкурентов?

Сохраните ответ в переменной average_calories и распечатайте переменную в терминале

In [113]:
average_calories = round(np.mean(calorie_stats), 2)
print(average_calories)

106.88


3. Корректно ли среднее количество калорий отражает распределение набора данных? Давайте отсортируем данные и посмотрим.

Отсортируйте данные и сохраните результат в переменной calorie_stats_sorted. Распечатайте отсортированную информацию

In [114]:
calorie_stats_sorted = np.sort(calorie_stats)
print(calorie_stats_sorted)

[ 50.  50.  50.  70.  70.  80.  90.  90.  90.  90.  90.  90.  90. 100.
 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100. 100.
 100. 100. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110. 110.
 110. 110. 110. 120. 120. 120. 120. 120. 120. 120. 120. 120. 120. 130.
 130. 140. 140. 140. 150. 150. 160.]


4. Похоже, что большинство значений выше среднего. Давайте посмотрим, является ли медиана наиболее корректным показателем набора данных.

Вычислите медиану набора данных и сохраните свой ответ в median_calories. Выведите медиану, чтобы вы могли видеть, как она сравнивается со средним значением.

In [115]:
median_calories = round(np.median(calorie_stats), 2)

print(f"Среднее значение: {average_calories}")
print(f"Медиана: {median_calories}")

Среднее значение: 106.88
Медиана: 110.0


5. В то время как медиана показывает, что по крайней мере половина наших значений составляет более 100 калорий, было бы более впечатляюще показать, что значительная часть конкурентов имеет более высокое количество калорий, чем CrunchieMunchies.

Рассчитайте различные процентили и распечатайте их, пока не найдете наименьший процентиль, превышающий 60 калорий. Сохраните это значение в переменной nth_percentile.

In [116]:
nth_percentile = np.percentile(calorie_stats,
                                np.where(np.percentile(calorie_stats, range(101)) > 60)[0][0])

print(f"Наименьший процентиль, превышающий 60 калорий: {nth_percentile}")

Наименьший процентиль, превышающий 60 калорий: 70.0


6. Хотя процентиль показывает нам, что у большинства конкурентов количество калорий намного выше, это неудобная концепция для использования в маркетинговых материалах.

Вместо этого давайте подсчитаем процент хлопьев, в которых содержится более 60 калорий на порцию. Сохраните свой ответ в переменной more_calories и распечатайте его

In [117]:
more_calories = (calorie_stats > 60).sum() / len(calorie_stats) * 100
print(f"Процент конкурентов с более чем 60 калориями: {more_calories:.1f}%")

Процент конкурентов с более чем 60 калориями: 96.1%


7. Это действительно высокий процент. Это будет очень полезно, когда мы будем продвигать CrunchieMunchies. Но один вопрос заключается в том, насколько велики различия в наборе данных? Можем ли мы сделать обобщение, что в большинстве злаков содержится около 100 калорий или разброс еще больше?

Рассчитайте величину отклонения, найдя стандартное отклонение, Сохраните свой ответ в calorie_std и распечатайте на терминале. Как мы можем включить эту ценность в наш анализ?

In [118]:
calorie_std = np.std(calorie_stats)

print(f"Стандартное отклонение: {calorie_std:.2f} калорий")
print(f"Среднее значение: {np.mean(calorie_stats):.2f} калорий")
print(f"Медиана: {np.median(calorie_stats):.2f} калорий")
print(f"\nВывод: Большое стандартное отклонение ({calorie_std:.2f}) показывает,")
print("что калорийность конкурентов сильно варьируется - от 50 до 160 калорий.")
print("CrunchieMunchies с 60 калориями находится в нижней части распределения,")
print("что делает их одним из самых низкокалорийных вариантов на рынке.")

Стандартное отклонение: 19.36 калорий
Среднее значение: 106.88 калорий
Медиана: 110.00 калорий

Вывод: Большое стандартное отклонение (19.36) показывает,
что калорийность конкурентов сильно варьируется - от 50 до 160 калорий.
CrunchieMunchies с 60 калориями находится в нижней части распределения,
что делает их одним из самых низкокалорийных вариантов на рынке.


8. Напишите короткий абзац, в котором кратко изложите свои выводы и то, как, по вашему мнению, эти данные могут быть использованы в интересах Mycrunch при маркетинге CrunchieMunchies.

Анализ данных конкурентов показывает, что CrunchieMunchies с 60 калориями на порцию имеют значительное преимущество перед большинством злаков на рынке. 96% конкурентов содержат более 60 калорий, что делает наш продукт одним из самых низкокалорийных вариантов. Средняя калорийность конкурентов составляет 107 калорий — почти вдвое выше, чем у CrunchieMunchies. При этом стандартное отклонение в 19 калорий указывает на широкий разброс значений, но даже минимальные значения у конкурентов (50 калорий) лишь незначительно ниже нашего показателя.

Маркетинговая стратегия:

Использовать слоган: "CrunchieMunchies — на 97% полезнее других злаков!"

Визуально показать разницу между 60 калориями и средними 107 калориями конкурентов


